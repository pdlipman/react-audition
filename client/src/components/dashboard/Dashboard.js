import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import ImmutablePropTypes from 'react-immutable-proptypes';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import TextField from 'material-ui/TextField';
import ClassSubject from './ClassSubject';

import {
  getStudentData,
} from './dashboardThunks';

const mapStateToProps = state => ({
  error: state.getIn(['dashboard', 'error']),
  classes: state.getIn(['dashboard', 'classes']),
});

const mapDispatchToProps = dispatch => ({
  getStudents: bindActionCreators(getStudentData, dispatch),
});

export class Dashboard extends Component {
  static propTypes = {
    classes: ImmutablePropTypes.map.isRequired,
    error: PropTypes.string,
    getStudents: PropTypes.func.isRequired,
  };

  static defaultProps = {
    classes: Map(),
    error: '',
  };

  componentWillMount() {
    const {
      getStudents,
    } = this.props;

    getStudents();
  }

  renderClasses() {
    const {
      classes,
    } = this.props;
    return classes.map((classSubject) => {
      return (
        <ClassSubject
          key={ classSubject.get('id') }
          id={ classSubject.get('id') }
          label={ classSubject.get('label') }
        />
      )
    }).toArray();
  }

  renderError() {
    const {
      error,
    } = this.props;
    return error && (
      <TextField
        error
        id='error'
        label='Error'
        defaultValue={ error }
        margin='normal'
      />
    );
  }

  render() {
    return (
      <div>
        { this.renderError() }
        { this.renderClasses() }
      </div>
    );
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard);