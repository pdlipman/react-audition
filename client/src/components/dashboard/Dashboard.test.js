import React from 'react';
import Immutable from 'immutable';

import { shallow } from 'enzyme';
import { Dashboard } from './Dashboard';

const classes = Immutable.fromJS({
  Math: {
    label: 'Math',
    id: 'Math',
  },
});

describe('Dashboard', () => {
  test('Dashboard should render properly', () => {
    const wrapper = shallow(
      <Dashboard
        classes={ classes }
        getStudents={ () => {} }
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('Dashboard should render with error properly', () => {
    const wrapper = shallow(
      <Dashboard
        classes={ classes }
        error={ 'Test Error Message'}
        getStudents={ () => {} }
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});