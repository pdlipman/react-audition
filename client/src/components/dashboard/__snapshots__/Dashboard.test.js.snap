// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Dashboard should render properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dashboard
    classes={
        Immutable.Map {
            Math: Immutable.Map {
                label: "Math",
                id: "Math",
            },
        }
    }
    error=""
    getStudents={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        "",
        Array [
          <Connect(ClassSubject)
            id="Math"
            label="Math"
/>,
        ],
      ],
    },
    "ref": null,
    "rendered": Array [
      "",
      Object {
        "instance": null,
        "key": "Math",
        "nodeType": "class",
        "props": Object {
          "id": "Math",
          "label": "Math",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "",
          Array [
            <Connect(ClassSubject)
              id="Math"
              label="Math"
/>,
          ],
        ],
      },
      "ref": null,
      "rendered": Array [
        "",
        Object {
          "instance": null,
          "key": "Math",
          "nodeType": "class",
          "props": Object {
            "id": "Math",
            "label": "Math",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Dashboard Dashboard should render with error properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dashboard
    classes={
        Immutable.Map {
            Math: Immutable.Map {
                label: "Math",
                id: "Math",
            },
        }
    }
    error="Test Error Message"
    getStudents={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <TextField
          defaultValue="Test Error Message"
          error={true}
          id="error"
          label="Error"
          margin="normal"
          required={false}
          select={false}
/>,
        Array [
          <Connect(ClassSubject)
            id="Math"
            label="Math"
/>,
        ],
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "defaultValue": "Test Error Message",
          "error": true,
          "id": "error",
          "label": "Error",
          "margin": "normal",
          "required": false,
          "select": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "Math",
        "nodeType": "class",
        "props": Object {
          "id": "Math",
          "label": "Math",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <TextField
            defaultValue="Test Error Message"
            error={true}
            id="error"
            label="Error"
            margin="normal"
            required={false}
            select={false}
/>,
          Array [
            <Connect(ClassSubject)
              id="Math"
              label="Math"
/>,
          ],
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "defaultValue": "Test Error Message",
            "error": true,
            "id": "error",
            "label": "Error",
            "margin": "normal",
            "required": false,
            "select": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "Math",
          "nodeType": "class",
          "props": Object {
            "id": "Math",
            "label": "Math",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
